File: apps/backend-api/src/lib/runtimeProfile.ts
import type { RuntimeProfile, RuntimeProfileName, RuntimeStrictness } from "@r3mes/shared-types";

const DEFAULT_QDRANT_COLLECTION = "r3mes_knowledge";
const DEFAULT_QDRANT_VECTOR_SIZE = 1024;

function normalizeProfileName(value: string | undefined, env: Record<string, string | undefined>): RuntimeProfileName {
  const raw = value?.trim().toLowerCase();
  if (raw === "local-dev" || raw === "eval" || raw === "pilot-rag" || raw === "production" || raw === "peft-lab") {
    return raw;
  }
  return env.NODE_ENV === "production" ? "production" : "local-dev";
}

function strictnessForProfile(name: RuntimeProfileName, env: Record<string, string | undefined>): RuntimeStrictness {
  if (
    name === "eval" ||
    name === "pilot-rag" ||
    name === "production" ||
    env.R3MES_REQUIRE_REAL_EMBEDDINGS === "1" ||
    env.R3MES_REQUIRE_REAL_RERANKER === "1"
  ) {
    return "quality_fallback_blocked";
  }
  return "dev_fallback_allowed";
}

function normalizeChatRuntime(value: string | undefined): RuntimeProfile["chat"]["runtime"] {
  return value?.trim().toLowerCase() === "transformers_peft" ? "transformers_peft" : "llama_cpp";
}

function normalizeEmbeddingProvider(value: string | undefined): RuntimeProfile["embedding"]["requestedProvider"] {
  const raw = value?.trim().toLowerCase();
  if (raw === "ai-engine" || raw === "ai_engine") return "ai-engine";
  if (raw === "bge-m3") return "bge-m3";
  return "deterministic";
}

function normalizeRerankerMode(value: string | undefined): RuntimeProfile["reranker"]["requestedMode"] {
  return value?.trim().toLowerCase() === "deterministic" ? "deterministic" : "model";
}

function positiveInt(value: string | undefined, fallback?: number): number | undefined {
  const parsed = Number.parseInt(value ?? "", 10);
  return Number.isFinite(parsed) && parsed > 0 ? parsed : fallback;
}

function qdrantRequired(env: Record<string, string | undefined>, realProvidersRequired: boolean): boolean {
  const retrievalEngine = env.R3MES_RETRIEVAL_ENGINE?.trim().toLowerCase();
  return realProvidersRequired || retrievalEngine === "qdrant" || retrievalEngine === "hybrid";
}

export function resolveRuntimeProfile(env: Record<string, string | undefined> = process.env): RuntimeProfile {
  const name = normalizeProfileName(env.R3MES_RUNTIME_PROFILE, env);
  const strictness = strictnessForProfile(name, env);
  const realProvidersRequired = strictness === "quality_fallback_blocked";
  const vectorSize = positiveInt(env.R3MES_QDRANT_VECTOR_SIZE, DEFAULT_QDRANT_VECTOR_SIZE) ?? DEFAULT_QDRANT_VECTOR_SIZE;

  return {
    version: 1,
    name,
    strictness,
    chat: {
      runtime: normalizeChatRuntime(env.R3MES_CHAT_RUNTIME ?? env.R3MES_AI_RUNTIME ?? env.R3MES_INFERENCE_BACKEND),
      modelFamily: env.R3MES_MODEL_FAMILY === "unknown" ? "unknown" : "qwen2_5_3b",
      modelId: env.R3MES_MODEL_ID ?? env.R3MES_CHAT_MODEL ?? "qwen2.5-3b",
      synthesisOnly: env.R3MES_CHAT_SYNTHESIS_ONLY !== "0",
      allowDeterministicComposerBypass: !realProvidersRequired,
    },
    embedding: {
      requestedProvider: normalizeEmbeddingProvider(env.R3MES_EMBEDDING_PROVIDER),
      requiredRealProvider: realProvidersRequired,
      expectedModelIncludes: ["bge-m3"],
      expectedDimension: vectorSize,
    },
    reranker: {
      requestedMode: normalizeRerankerMode(env.R3MES_RERANKER_MODE),
      requiredRealProvider: realProvidersRequired,
      expectedProvider: "cross_encoder",
    },
    qdrant: {
      required: qdrantRequired(env, realProvidersRequired),
      collectionName: env.R3MES_QDRANT_COLLECTION ?? DEFAULT_QDRANT_COLLECTION,
      vectorSize,
    },
    stream: {
      productMode: env.R3MES_STREAM_PRODUCT_MODE === "sse_stream" ? "sse_stream" : "non_stream_json",
    },
    lora: {
      role: "behavior_persona_only",
      optional: true,
      maxLockWaitMs: positiveInt(env.R3MES_LORA_MAX_LOCK_WAIT_MS),
    },
  };
}

File: apps/ai-engine/r3mes_ai_engine/runtime_profile.py
from __future__ import annotations

from typing import Any, Literal

from r3mes_ai_engine.settings import Settings

RuntimeProfileName = Literal["local-dev", "eval", "pilot-rag", "production", "peft-lab"]
RuntimeStrictness = Literal["dev_fallback_allowed", "quality_fallback_blocked"]

_STRICT_PROFILES: set[str] = {"eval", "pilot-rag", "production"}


def runtime_strictness(profile_name: str) -> RuntimeStrictness:
    if profile_name in _STRICT_PROFILES:
        return "quality_fallback_blocked"
    return "dev_fallback_allowed"


def is_strict_runtime_profile(settings: Settings) -> bool:
    return runtime_strictness(settings.runtime_profile) == "quality_fallback_blocked"


def runtime_profile_summary(settings: Settings) -> dict[str, Any]:
    """Return the ai-engine mirror of the product runtime profile.

    This is intentionally a summary: backend owns retrieval/Qdrant policy, while
    ai-engine reports the runtime facts it can enforce locally.
    """

    embedding_provider = (
        "bge-m3" if "bge-m3" in settings.embedding_model_name_or_path.lower() else "ai-engine"
    )
    strictness = runtime_strictness(settings.runtime_profile)
    return {
        "version": 1,
        "name": settings.runtime_profile,
        "strictness": strictness,
        "chat": {
            "runtime": settings.inference_backend,
            "model_family": "qwen2_5_3b"
            if "qwen2.5-3b" in settings.default_model_name.lower()
            or "qwen2.5-3b" in settings.hf_model_name_or_path.lower()
            else "unknown",
            "model_id": settings.default_model_name
            if settings.inference_backend == "llama_cpp"
            else settings.hf_model_name_or_path,
            "synthesis_only": True,
            "allow_deterministic_composer_bypass": settings.runtime_profile == "local-dev",
        },
        "embedding": {
            "requested_provider": embedding_provider,
            "required_real_provider": strictness == "quality_fallback_blocked",
            "expected_model_includes": ["bge-m3"],
            "expected_dimension": 1024,
        },
        "reranker": {
            "requested_mode": "model",
            "required_real_provider": strictness == "quality_fallback_blocked",
            "expected_provider": "cross_encoder",
        },
        "stream": {
            "product_mode": "sse_stream",
            "diagnostics": "first_event_and_headers",
        },
        "lora": {
            "role": "behavior_persona_only",
            "optional": True,
            "max_lock_wait_ms": settings.lora_max_lock_wait_ms,
            "budget_policy": "fail_request"
            if strictness == "quality_fallback_blocked"
            else "warn_continue",
        },
    }
